File: Functions/insertNaNAtGaps.py
"""
insertNaNAtGaps - Insert NaN values at data gaps

This function detects and inserts NaN values at data gaps to prevent
continuous lines in plots across discontinuous data segments.
"""

import numpy as np
import pandas as pd
import time


def insert_nan_at_gaps(time_yy_mm_dd, dwell_time_s, voltage_v, current_a, cell_temp_c, chamber_temp_c):
    """
    Insert NaN values at data gaps.
    
    Parameters
    ----------
    time_yy_mm_dd : numpy.ndarray
        Datetime array of timestamps (datetime64)
    dwell_time_s : numpy.ndarray
        Dwell time in seconds
    voltage_v : numpy.ndarray
        Cell voltage in Volts
    current_a : numpy.ndarray
        Cell current in Amperes
    cell_temp_c : numpy.ndarray
        Cell surface temperature in Celsius
    chamber_temp_c : numpy.ndarray
        Chamber ambient temperature in Celsius
    
    Returns
    -------
    time_with_gaps : numpy.ndarray
        Datetime array with NaT inserted at gaps
    time_s : numpy.ndarray
        Time in seconds with NaN at gaps
    voltage : numpy.ndarray
        Voltage with NaN at gaps
    current : numpy.ndarray
        Current with NaN at gaps
    cell_temp : numpy.ndarray
        Cell temperature with NaN at gaps
    chamber_temp : numpy.ndarray
        Chamber temperature with NaN at gaps
    """
    
    print('Detecting and processing data gaps...')
    start_time = time.time()
    
    # Calculate the time differences between consecutive data points
    time_diff = np.diff(pd.to_datetime(time_yy_mm_dd))
    
    # Find indices where the gap is longer than 1 minute
    gap_indices = np.where(time_diff > pd.Timedelta(minutes=1))[0] + 1  # +1 to get position after gap
    gap_indices = gap_indices.tolist()
    
    # Find periods where current is zero for certain amount of samples and voltage lower than threshold
    voltage_threshold = 3.3  # minimum voltage with zero current
    rest_samples = 60  # minimum number of samples to detect with 0 current
    count_time_samples = 0
    current_state = 1
    
    for i in range(len(dwell_time_s)):
        if (current_a[i] == 0) and (voltage_v[i] < voltage_threshold):
            count_time_samples += 1
            if (count_time_samples >= rest_samples) and current_state:
                gap_indices.append(i)
                current_state = 0
        else:
            count_time_samples = 0
            current_state = 1
    
    gap_indices = sorted(gap_indices)
    print(f'Found {len(gap_indices)} data gaps. Inserting NaN values...')
    start_time = time.time()
    
    # Pre-allocate arrays with final size (much faster than repeated concatenation)
    num_gaps = len(gap_indices)
    original_length = len(time_yy_mm_dd)
    final_length = original_length + num_gaps
    
    # Pre-allocate with appropriate types - match the dtype of the input time array
    time_with_gaps = np.full(final_length, np.datetime64('NaT'), dtype=time_yy_mm_dd.dtype)
    dwell_time_with_gaps = np.full(final_length, np.nan)
    voltage_with_gaps = np.full(final_length, np.nan)
    current_with_gaps = np.full(final_length, np.nan)
    cell_temp_with_gaps = np.full(final_length, np.nan)
    chamber_temp_with_gaps = np.full(final_length, np.nan)
    
    # Fill in the data with gaps
    source_idx = 0
    dest_idx = 0
    
    for i, gap_pos in enumerate(gap_indices):
        # Copy data up to the gap
        chunk_size = gap_pos - source_idx
        if chunk_size > 0:
            time_with_gaps[dest_idx:dest_idx+chunk_size] = time_yy_mm_dd[source_idx:gap_pos]
            dwell_time_with_gaps[dest_idx:dest_idx+chunk_size] = dwell_time_s[source_idx:gap_pos]
            voltage_with_gaps[dest_idx:dest_idx+chunk_size] = voltage_v[source_idx:gap_pos]
            current_with_gaps[dest_idx:dest_idx+chunk_size] = current_a[source_idx:gap_pos]
            cell_temp_with_gaps[dest_idx:dest_idx+chunk_size] = cell_temp_c[source_idx:gap_pos]
            chamber_temp_with_gaps[dest_idx:dest_idx+chunk_size] = chamber_temp_c[source_idx:gap_pos]
            dest_idx += chunk_size
        
        # NaN is already in place from pre-allocation
        dest_idx += 1  # Skip the NaN position
        source_idx = gap_pos
        
        # Show progress every 5%
        if i % max(1, round(num_gaps * 0.05)) == 0:
            percent_complete = round(i / num_gaps * 100)
            print(f'  Progress: {percent_complete}% ({i}/{num_gaps} gaps processed)')
    
    # Copy remaining data after last gap
    if source_idx < original_length:
        remaining_size = original_length - source_idx
        time_with_gaps[dest_idx:dest_idx+remaining_size] = time_yy_mm_dd[source_idx:]
        dwell_time_with_gaps[dest_idx:dest_idx+remaining_size] = dwell_time_s[source_idx:]
        voltage_with_gaps[dest_idx:dest_idx+remaining_size] = voltage_v[source_idx:]
        current_with_gaps[dest_idx:dest_idx+remaining_size] = current_a[source_idx:]
        cell_temp_with_gaps[dest_idx:dest_idx+remaining_size] = cell_temp_c[source_idx:]
        chamber_temp_with_gaps[dest_idx:dest_idx+remaining_size] = chamber_temp_c[source_idx:]
    
    print(f'NaN insertion complete. (Elapsed: {time.time() - start_time:.2f} s)')
    
    # Assign to output variable names
    time_s = dwell_time_with_gaps
    voltage = voltage_with_gaps
    current = current_with_gaps
    cell_temp = cell_temp_with_gaps
    chamber_temp = chamber_temp_with_gaps
    
    return time_with_gaps, time_s, voltage, current, cell_temp, chamber_temp

File: Functions/insertNaNAtGaps.m
function [timeWithGaps, timeS, voltage, current, cellTemp, chamberTemp] = ...
    insertNaNAtGaps(timeYYMMDD, dwellTimeS, voltageV, currentA, cellTempC, chamberTempC)
% insertNaNAtGaps - Insert NaN values at data gaps
%
% This function detects and inserts NaN values at data gaps to prevent
% continuous lines in plots across discontinuous data segments.
%
% Inputs:
%   timeYYMMDD   - Datetime array of timestamps
%   dwellTimeS   - Dwell time in seconds
%   voltageV     - Cell voltage in Volts
%   currentA     - Cell current in Amperes
%   cellTempC    - Cell surface temperature in Celsius
%   chamberTempC - Chamber ambient temperature in Celsius
%
% Outputs:
%   timeWithGaps - Datetime array with NaN inserted at gaps
%   timeS        - Time in seconds with NaN at gaps
%   voltage      - Voltage with NaN at gaps
%   current      - Current with NaN at gaps
%   cellTemp     - Cell temperature with NaN at gaps
%   chamberTemp  - Chamber temperature with NaN at gaps

fprintf('Detecting and processing data gaps...\n');
tic;  % Restart timer

% Calculate the time differences between consecutive data points
timeDiff = diff(timeYYMMDD);

% Find indices where the gap is longer than 1 minute
gapIndices = find(timeDiff > minutes(1));

% Find periods where current is zero for certain amount of samples and voltage lower than threshold
VoltageThreshold = 3.3; % minimum voltage with zero current
RestSamples = 60; % minimum number of samples to detect with 0 current
CountTimeSamples=0;CurrentState=1;
for i=1:length(dwellTimeS)
    if (currentA(i)==0) && voltageV(i)<VoltageThreshold
        CountTimeSamples=CountTimeSamples+1;
        if (CountTimeSamples>=RestSamples) && CurrentState
            gapIndices = [gapIndices;i];
            CurrentState = 0;
        end
    else
        CountTimeSamples=0;CurrentState=1;
    end
end
gapIndices = sort(gapIndices);
fprintf('Found %d data gaps. Inserting NaN values...\n', length(gapIndices));
tic;  % Restart timer

% Pre-allocate arrays with final size (much faster than repeated concatenation)
numGaps = length(gapIndices);
originalLength = length(timeYYMMDD);
finalLength = originalLength + numGaps;

% Pre-allocate with appropriate types
timeWithGaps = NaT(finalLength, 1);
dwellTimeWithGaps = NaN(finalLength, 1);
voltageWithGaps = NaN(finalLength, 1);
currentWithGaps = NaN(finalLength, 1);
cellTempWithGaps = NaN(finalLength, 1);
chamberTempWithGaps = NaN(finalLength, 1);

% Fill in the data with gaps
sourceIdx = 1;
destIdx = 1;
for i = 1:numGaps
    gapPos = gapIndices(i);
    % Copy data up to the gap
    chunkSize = gapPos - sourceIdx + 1;
    if chunkSize > 0
        timeWithGaps(destIdx:destIdx+chunkSize-1) = timeYYMMDD(sourceIdx:gapPos);
        dwellTimeWithGaps(destIdx:destIdx+chunkSize-1) = dwellTimeS(sourceIdx:gapPos);
        voltageWithGaps(destIdx:destIdx+chunkSize-1) = voltageV(sourceIdx:gapPos);
        currentWithGaps(destIdx:destIdx+chunkSize-1) = currentA(sourceIdx:gapPos);
        cellTempWithGaps(destIdx:destIdx+chunkSize-1) = cellTempC(sourceIdx:gapPos);
        chamberTempWithGaps(destIdx:destIdx+chunkSize-1) = chamberTempC(sourceIdx:gapPos);
        destIdx = destIdx + chunkSize;
    end
    % NaN is already in place from pre-allocation
    destIdx = destIdx + 1;  % Skip the NaN position
    sourceIdx = gapPos + 1;
    
    % Show progress every 5%
    if mod(i, max(1, round(numGaps * 0.05))) == 0
        percentComplete = round(i / numGaps * 100);
        fprintf('  Progress: %d%% (%d/%d gaps processed)\n', percentComplete, i, numGaps);
    end
end

% Copy remaining data after last gap
if sourceIdx <= originalLength
    remainingSize = originalLength - sourceIdx + 1;
    timeWithGaps(destIdx:destIdx+remainingSize-1) = timeYYMMDD(sourceIdx:end);
    dwellTimeWithGaps(destIdx:destIdx+remainingSize-1) = dwellTimeS(sourceIdx:end);
    voltageWithGaps(destIdx:destIdx+remainingSize-1) = voltageV(sourceIdx:end);
    currentWithGaps(destIdx:destIdx+remainingSize-1) = currentA(sourceIdx:end);
    cellTempWithGaps(destIdx:destIdx+remainingSize-1) = cellTempC(sourceIdx:end);
    chamberTempWithGaps(destIdx:destIdx+remainingSize-1) = chamberTempC(sourceIdx:end);
end

fprintf('NaN insertion complete. (Elapsed: %.2f s)\n', toc);

% Assign to output variable names
timeS = dwellTimeWithGaps;
voltage = voltageWithGaps;
current = currentWithGaps;
cellTemp = cellTempWithGaps;
chamberTemp = chamberTempWithGaps;

end
